src/ast/irbuilderbpf.cpp

```diff
@@ -593,7 +593,8 @@ Value *IRBuilderBPF::CreateUSDTReadArgument(Value *ctx,
     {
       // Zero out `dst` in case we read less than 64 bits
       CreateStore(getInt64(0), dst);
-      CreateProbeRead(ctx, dst, abs_size, reg, as, loc);
+      // TODO : probe_read_user does not work here.
```

Comment:
Having a probe_read_kernel, works here. Not sure why.. (See usdt_semaphore test)

Reply:
Here, we read the value (`reg`) in `struct pt_regs`, so we need to use `probe_read_kernel()`. But it is a BPF context data so I think we can simply load the value, i.e.,

```cpp
{
       result = CreateLoad(reg);
       result = CreateIntCast(result, getInt64Ty(), false);
}
```

Reply:
:+1:  I will  use your load suggestion here.

Reply:
I tried the simple load here. But it didnt work for usdt_semphore_test - str(arg). However, reading reg using probe_read_kernel() did work.

Reply:
What errors do you have? Actually, in my environment (x64), the USDT arguments of usdt_semaphore_test are placed on a stack, so this if block are not executed.

```
% readelf --notes ./tests/testprogs/usdt_semaphore_test
[...]
Displaying notes found in: .note.stapsdt
  Owner                 Data size       Description
  stapsdt              0x00000047       NT_STAPSDT (SystemTap probe descriptors)
    Provider: tracetest
    Name: testprobe
    Location: 0x00000000004005d1, Base: 0x0000000000400687, Semaphore: 0x0000000000601038
    Arguments: -8@-136(%rbp) 8@-144(%rbp)
```

But the following program executes the block,  and simple load works.

```c
#include <stdio.h>
#include <sys/sdt.h>
void f(int a){
  DTRACE_PROBE1(foo, bar, a);
}
int main(){
    f(3);
    return 0;
}
```

```
% readelf --notes ./a.out
[...]
Displaying notes found in: .note.stapsdt
  Owner                 Data size       Description
  stapsdt              0x00000028       NT_STAPSDT (SystemTap probe descriptors)
    Provider: foo
    Name: bar
    Location: 0x0000000000000610, Base: 0x00000000000006a4, Semaphore: 0x0000000000000000
    Arguments: -4@%edi
```

```
% sudo ./src/bpftrace -e 'usdt:./a.out:foo:bar { printf("val: %d\n", arg0); }' -c ./a.out
Attaching 1 probe...
val: 3
```

```
Bytecode:
0: (bf) r6 = r1
1: (b7) r1 = 0
2: (7b) *(u64 *)(r10 -16) = r1
last_idx 2 first_idx 0
regs=2 stack=0 before 1: (b7) r1 = 0
3: (71) r1 = *(u8 *)(r6 +112)
[...]
```

Reply:
ah,  load instruciton should be `result = CreateLoad(getInt64Ty(), reg);`

Reply:
This works. Great. cant we just have signed IntCast here?

Reply:
Eventually we need to move this to the semantic analysis so that we can get rid of all these casts. I'm doing that in my 1091_integer_size branch, so for now this should be ok (but use `CreateInt()` instead of `SizedType()`)

Reply:
Ok. Will do.